src/d3a/models/strategy/area_agents/two_sided_pay_as_bid_engine.py

```diff
@@ -145,24 +146,36 @@ def event_bid_traded(self, *, bid_trade):
                               market_bid.buyer, market_bid.seller)
             assert bid_trade.offer.energy <= market_bid.energy, \
                 f"Traded bid on target market has more energy than the market bid."
-            self.markets.source.accept_bid(
+            source_trade = self.markets.source.accept_bid(
                 updated_bid,
                 energy=bid_trade.offer.energy,
                 seller=self.owner.name,
                 already_tracked=False,
                 trade_rate=updated_bid.price / updated_bid.energy
             )
-            self.delete_forwarded_bids(bid_info)
 
+            self.after_successful_trade_event(source_trade, bid_info)
         # Bid was traded in the source market by someone else
         elif bid_trade.offer.id == bid_info.source_bid.id:
-            if self.owner.name == bid_trade.seller:
-                return
             self.delete_forwarded_bids(bid_info)
         else:
             raise Exception(f"Invalid bid state for IAA {self.owner.name}: "
                             f"traded bid {bid_trade} was not in offered bids tuple {bid_info}")
 
+    def after_successful_trade_event(self, source_trade, bid_info):
```

Comment:
Would it be a good idea to move this method to IAAEngine class, and using it to handle both offers and bids?

Reply:
The logic is different in both cases. In the case of the offer the information about the residual offer has to be parsed before the accept_offer, because IIRC it needs to be associated with that. For the bids we do not need that, therefore the code here will actually be a bit different.